### ISSUE TITLE
`jax.random.normal` values depend on the number of shards: a draw appears fixed by (shard index, index within shard)

### ISSUE THREAD

**Issue by loofahcus** [jax-ml/jax#40232]
### Question

With `jax_threefry_partitionable=True` (the default; jax 0.8.3 here), is the *value* of a
`jax.random.normal` draw intended to be independent of how the output is sharded --
specifically, of how many shards the array is split into?

We observe that it is not, and we have measured the rule it appears to follow instead.
We would like to know whether that is a bug or an accepted consequence of the
partitionable implementation. We could not answer it from the documentation: discussion
#18480 states that the RNG "will remain deterministic", but discusses determinism with
respect to the key and the algorithm choice, and the config flag's help text mentions
only that communication overheads disappear. Sharding is not addressed either way.

### What we measured

An array is produced by `jax.random.normal` inside a `jit` whose `out_shardings` shard
one of its axes -- length 496 -- across a mesh axis. We vary only how many shards that
axis is cut into (4, 8, 16), holding the total device count (256), the key, and
everything else fixed. Rows per shard are then 124, 62, 31.

Reading single elements along that axis:

| comparison | rows/shard | first index whose value differs |
|---|---|---|
| 4 vs 8 shards | 124 vs 62 | **62** |
| 8 vs 16 shards | 62 vs 31 | **31** |

Below the crossover the values are bit-identical; from it onward they differ. In both
cases the crossover is exactly `axis_length / shard_count` -- the first row that lands in
shard 1 of the finer mesh while still being in shard 0 of the coarser one.

That suggested a rule, which we then stated as a prediction and tested rather than fitted:
**a draw is determined by (shard index, index within that shard), independent of its
global position.** If so, the same value must reappear at different global indices under
different shard counts, related by the ratio of shard widths:

```
f4[124*s + r] == f8[62*s + r] == f16[31*s + r]
```

Measured: **9 of 9** three-way triples matched, across shards 0 through 3, plus **12 of
12** pairs for the 8-vs-16 comparison sampled on its own. A value sitting at global index
`124` under 4 shards is the same value that sits at `62` under 8 shards and at `31` under
16.

Consistent with the same rule: arrays whose sharding does not include the varied axis are
bit-identical across all three meshes, and shard 0 is always bit-identical -- which is
what the rule predicts, since there the local and global indices coincide.

### Where it comes from in the HLO

The compiled computation selects each device's RNG offset by indexing a constant table
with `partition-id`, inside `jit(_uniform)`:

```
%partition-id.12 = u32[] partition-id()
%dynamic-slice.31 = s32[1] dynamic-slice(%constant.2789, %partition-id.12),
    metadata={op_name=".../jit(_normal)/jit(_normal_real)/jit(_uniform)/iota_2x32_shape"}
```

`%constant.2789` is `s32[256]`, and its contents are elided in both the before- and
after-optimization text dumps, so we could not read it directly. We measured the quantity
it encodes instead: shard a `broadcasted_iota` over the same mesh and read which global
index each addressable shard begins at.

| shard count | shard starts (first two, per process) |
|---|---|
| 4 | `[0, 64]` |
| 8 | `[0, 32]` |
| 16 | `[0, 16]` |

So the offsets are laid out by shard geometry. Combined with an `iota` generated over the
*local* shard shape, the identity "element i receives counter i" survives only where the
local offset coincides with the global index -- which is exactly what the crossover data
shows.

This is not about device placement. Holding the mesh *shape* fixed and reversing the
device order along the sharded axis gives bit-identical output, with a 6e-08
sum-of-squares wobble confirming the permutation was live. The table does compensate for
which physical device is which; what it does not compensate for is splitting the same
logical array into a different number of pieces.

### What we ruled out

Each measured on hardware, on a 256-device mesh, not argued:

- building optimizer state in the same `pjit` as the parameter init;
- the `prebuilt` argument to the initializer;
- how the `out_shardings` objects are constructed -- from a spec's `.sharding` property,
  or by hand from `PartitionSpec`, with or without `memory_kind`: three variants, outputs
  bit-identical to each other, and all three diverging across meshes;
- the device order of the mesh (identical to a freshly built one);
- XLA environment, including a TPU compiler-behaviour override and premapped-buffer
  settings: not one bit changed;
- graph structure: RNG-related `partition-id` appears exactly twice in both the diverging
  and non-diverging paths, under `_uniform/iota_2x32_shape`, in the same table form.

### On minimisation

We could not reduce this to a small standalone script. Roughly a dozen attempts -- pure
JAX with the same shapes and PartitionSpecs, at 4, 8 and 256 devices, single- and
multi-process, up to 60 arrays drawn in one sharded `jit` -- were all sharding-invariant.

The honest reason is that those scripts computed *different values* from the program that
diverges, so they were never running the same computation and their invariance is not
evidence. What does reproduce it is the real initializer under the real mesh, and every
measurement above was taken there.

If it would help, we are glad to try a specific construction the maintainers expect to
hit the partitioned-RNG path, or to run a patched JAX that prints the offset table -- the
latter would settle the mechanism outright, since the table contents are the one thing we
could not read.

### Why it matters

The effect is silent. Two training runs of one configuration that differ only in the
device mesh begin from unrelated initial weights, and nothing in the loss makes that
visible -- the curves simply differ a little. Every other explanation (precision,
optimizer, data pipeline, checkpointing) is excluded by the weights already differing
before the first step, but arriving at that conclusion is slow precisely because nothing
points at the RNG.

If value-invariance is not intended here, we think it is worth saying so explicitly in
the flag's documentation or in #18480, since a reader currently has no reason to expect
the values to move with the mesh.

### Related, but not the same

Discussion #18480 announced this algorithm change and states the RNG "will remain
deterministic", but only with respect to the key and the algorithm -- device count and
sharding are not mentioned. #26019 (closed) also concerns the partitionable Threefry, but
is about sample quality under KS tests rather than values moving with the sharding.

### Environment

- jax / jaxlib 0.8.3, TPU v6e, 256 devices, 64 processes
- `jax_threefry_partitionable` at its default of True
- `jax_enable_x64` False

### What would resolve this

1. Whether values are meant to be sharding-invariant under this flag.
2. If yes, a fix; if no, a documentation note plus the supported way to obtain a
   sharding-invariant draw for parameter initialization.
3. Either way, a pointer to how to force the partitioned-RNG path in a small program, so
   this can be covered by a regression test rather than by a 256-device run.

### System info (python version, jaxlib version, accelerator, etc.)

```
jax:    0.8.3
jaxlib: 0.8.3
numpy:  2.1.3
python: 3.12.12 (main, Oct 31 2025, 23:02:31) [Clang 21.1.4 ]
device info: TPU v6 lite-256, 4 local devices"
process_count: 64
platform: uname_result(system='Linux', node='<redacted>', release='6.6.113+',
                       version='#1 SMP Sat Nov 29 10:43:19 UTC 2025', machine='x86_64')
```

**Comment by jakevdp** [comment]
You mention running on TPU – is it possible that you've somewhere enabled the `rbg` or `unsafe_rbg` implementation? If so, producing different values for different shardings would be expected, as the rbg impl on TPU is not batch-invariant or shard-invariant.

**Comment by loofahcus** [comment]
@jakevdp You are right, and that invalidates the premise. Our framework (AXLearn) sets both of
these before any of our code runs:
```python
# axlearn/common/utils_spmd.py -- unconditional, during distributed setup
jax.config.update("jax_default_prng_impl", "rbg")
# axlearn/common/compiler_options.py -- into LIBTPU_INIT_ARGS before importing jax
xla_tpu_spmd_rng_bit_generator_unsafe=True,
```
That puts us in the `rbg**` column of the summary table in the `jax.random` docstring --
the one where "identical across shardings" is blank. So everything we measured is
documented behaviour, and `jax_threefry_partitionable=True` was accurate but irrelevant,
since the threefry generation path was never taken. I had also drafted a suggestion that
the docs should mention this; checking before posting, they already do, clearly.
It explains our "On minimisation" loose end too: the standalone reproducers produced a
third set of values because they never called the framework's setup, so they ran threefry
without the XLA flag -- a different algorithm entirely.
The check we never ran: print `jax.config.jax_default_prng_impl`. Apologies for the noise,
and thanks for the fast and exactly-correct diagnosis.
Closing. Whether that rbg choice is still warranted now that threefry is partitionable is
a question for our side, not JAX's.

**Comment by jakevdp** [comment]
Great! Glad the issue is resolved.
